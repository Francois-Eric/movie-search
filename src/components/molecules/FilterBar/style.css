.FilterBar {
  margin: 2rem auto;
}

.FilterBar--multiple {
  display: grid;
  gap: 2rem;
  grid-template-columns: 1fr 1fr 1fr;
}

.FilterBar--single {
  width: 50%;
}

.FilterBar__toggle {
  display: none;
}
