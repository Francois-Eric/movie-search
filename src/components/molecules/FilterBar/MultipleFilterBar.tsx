import { Autocomplete, TextField } from "@mui/material";
import { Dispatch, SetStateAction, useState } from "react";
import CONFIG from "../../../config.json";
import { FilmLocationFilters } from "../../../types";
import "./style.css";

interface Props {
  filters: FilmLocationFilters[];
  selectedFilters: { [key: string]: string[] };
  setSelectedFilters: Dispatch<SetStateAction<{ [key: string]: string[] }>>;
}

export const MultipleFilterBar = ({
  filters,
  selectedFilters,
  setSelectedFilters,
}: Props) => {
  const filterableFields = CONFIG.api.filmLocation.fields.reduce(
    (previousValue, currentValue) =>
      currentValue.filterable
        ? [...previousValue, currentValue]
        : previousValue,
    [] as { id: string; label: string }[]
  );

  return (
    <div className={"FilterBar FilterBar--multiple"}>
      {filterableFields.map(({ id, label }) => {
        const options =
          filters
            ?.find((filter) => filter.id === id)
            ?.values.map((value) => value.name)
            .sort() || [];
        return (
          <Autocomplete
            multiple
            key={id}
            id={id}
            options={options}
            onChange={(event: any, newValue: string[]) => {
              setSelectedFilters({ ...selectedFilters, [id]: newValue });
            }}
            renderInput={(params) => <TextField {...params} label={label} />}
          />
        );
      })}
    </div>
  );
};
