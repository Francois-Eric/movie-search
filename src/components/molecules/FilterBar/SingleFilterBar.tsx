import { Autocomplete, TextField } from "@mui/material";
import { Dispatch, SetStateAction, useEffect, useState } from "react";
import CONFIG from "../../../config.json";
import { FilmLocationFilters } from "../../../types";
import "./style.css";

interface Props {
  filters: FilmLocationFilters[];
  setSelectedFilters: Dispatch<SetStateAction<{ [key: string]: string[] }>>;
}

export const SingleFilterBar = ({ filters, setSelectedFilters }: Props) => {
  const [options, setOptions] = useState<{ [key: string]: string[] }>({});
  const filterableFields = CONFIG.api.filmLocation.fields.reduce(
    (previousValue, currentValue) =>
      currentValue.filterable
        ? [...previousValue, currentValue]
        : previousValue,
    [] as { id: string; label: string }[]
  );

  const handleChange = (newValue: string[]) => {
    const newFilters: { [key: string]: string[] } = {};
    newValue.forEach((val) => {
      options[val]?.forEach((filter: string) => {
        if (newFilters[filter]) {
          if (newFilters[filter].indexOf(val) === -1) {
            newFilters[filter].push(val);
          }
        } else {
          newFilters[filter] = [val];
        }
      });
    });
    setSelectedFilters(newFilters);
  };

  useEffect(() => {
    const formatedOptions: { [key: string]: string[] } = {};

    filterableFields.forEach(({ id }) => {
      filters
        .find((filter) => filter.id === id)
        ?.values.forEach(({ name }) => {
          if (formatedOptions[name]) {
            if (formatedOptions[name].indexOf(id) === -1) {
              formatedOptions[name].push(id);
            }
          } else {
            formatedOptions[name] = [id];
          }
        });
    });

    setOptions(formatedOptions);
  }, []);

  return (
    <div className="FilterBar FilterBar--single">
      <Autocomplete
        multiple
        id={filterableFields.map((field) => field.label).join(", ")}
        options={Object.keys(options).sort()}
        onChange={(event: any, newValue: string[]) => handleChange(newValue)}
        renderInput={(params) => (
          <TextField
            {...params}
            label={filterableFields.map((field) => field.label).join(", ")}
          />
        )}
      />
    </div>
  );
};
