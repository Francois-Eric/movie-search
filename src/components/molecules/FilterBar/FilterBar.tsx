import { Switch } from "@mui/material";
import { Dispatch, SetStateAction, useState } from "react";
import { FilmLocationFilters } from "../../../types";
import { MultipleFilterBar } from "./MultipleFilterBar";
import { SingleFilterBar } from "./SingleFilterBar";

interface Props {
  filters: FilmLocationFilters[];
  selectedFilters: { [key: string]: string[] };
  setSelectedFilters: Dispatch<SetStateAction<{ [key: string]: string[] }>>;
}

export const FilterBar = ({
  filters,
  selectedFilters,
  setSelectedFilters,
}: Props) => {
  const [isMultipleFilterBar, setIsMultipleFilterBar] = useState(false);

  return (
    <>
      <div className="FilterBar__toggle">
        <Switch
          checked={isMultipleFilterBar}
          onChange={() => {
            setIsMultipleFilterBar(!isMultipleFilterBar);
            setSelectedFilters({});
          }}
          inputProps={{ "aria-label": "controlled" }}
        />
      </div>
      {isMultipleFilterBar ? (
        <MultipleFilterBar
          filters={filters}
          selectedFilters={selectedFilters}
          setSelectedFilters={setSelectedFilters}
        />
      ) : (
        <SingleFilterBar
          filters={filters}
          setSelectedFilters={setSelectedFilters}
        />
      )}
    </>
  );
};
