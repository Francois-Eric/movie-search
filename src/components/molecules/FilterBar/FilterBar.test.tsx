import { render } from "@testing-library/react";
import {
  mockedFilmLocationFilters,
  mockedFilmLocationSelectedFilters,
} from "../../../mocks/FilmLocationMocks";
import { FilterBar } from "./FilterBar";

describe("<FilterBar/>", () => {
  test("should render correctly", () => {
    const { container } = render(
      <FilterBar
        filters={mockedFilmLocationFilters}
        selectedFilters={mockedFilmLocationSelectedFilters}
        setSelectedFilters={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
