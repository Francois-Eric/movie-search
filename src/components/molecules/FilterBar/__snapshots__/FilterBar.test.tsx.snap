// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterBar/> should render correctly 1`] = `
<div>
  <div
    class="FilterBar__toggle"
  >
    <span
      class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
    >
      <span
        class="MuiSwitch-switchBase MuiSwitch-colorPrimary MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      >
        <input
          aria-label="controlled"
          class="MuiSwitch-input PrivateSwitchBase-input css-1m9pwf3"
          type="checkbox"
        />
        <span
          class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
      />
    </span>
  </div>
  <div
    class="FilterBar FilterBar--single"
  >
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="Année, Type, Code Postal"
          id="Année, Type, Code Postal-label"
        >
          Année, Type, Code Postal
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="Année, Type, Code Postal"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Année, Type, Code Postal
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
